miio/gateway.py

```diff
@@ -529,45 +581,295 @@ def set_light(self, color_name, brightness):
-            self.type,
+            self.device_type,
             self.sid,
             self.get_firmware_version(),
-            self.battery(),
+            self.get_battery(),
         )
 
+    @property
+    def device_type(self):
+        """Return the device type name."""
+        return self.type.name
+
+    @property
+    def battery(self):
+        """Return the battery level."""
+        return self._battery
+
+    @command()
+    def subdevice_send(self, command):
+        """Send a command/query to the subdevice"""
+        try:
+            return self._gw.send(command, [self.sid])
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while sending command %s" % (command)
+            ) from ex
+            return None
+
+    @command()
+    def subdevice_send_arg(self, command, arguments):
+        """Send a command/query including arguments to the subdevice"""
+        try:
+            return self._gw.send(command, arguments, extra_parameters={"sid": self.sid})
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while sending "
+                "command '%s' with arguments '%s'" % (command, str(arguments))
+            ) from ex
+
+    @command(click.argument("property"))
+    def get_subdevice_prop(self, property):
+        """Get the value of a property of the subdevice."""
+        try:
+            response = self._gw.send("get_device_prop", [self.sid, property])
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while fetching property %s" % (property)
+            ) from ex
+
+        if not response:
+            raise GatewayException(
+                "Empty response while fetching property %s: %s" % (property, response)
+            )
+
+        return response
+
+    @command(click.argument("properties", nargs=-1))
+    def get_subdevice_prop_exp(self, properties):
+        """Get the value of a bunch of properties of the subdevice."""
+        try:
+            response = self._gw.send(
+                "get_device_prop_exp", [[self.sid] + list(properties)]
+            ).pop()
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while fetching properties %s: %s" % (properties)
+            ) from ex
+
+        if len(list(properties)) != len(response):
+            raise GatewayException(
+                "unexpected result while fetching properties %s: %s"
+                % (properties, response)
+            )
+
+        for item in response:
+            if not item:
+                raise GatewayException(
+                    "One or more empty results while "
+                    "fetching properties %s: %s" % (properties, response)
+                )
+
+        return response
+
+    @command(click.argument("property"), click.argument("value"))
+    def set_subdevice_prop(self, property, value):
```

Comment:
`set_prop`? Or maybe all getters/setters should be using non-shortened form for prop actually... `set_property`, `get_property` are not that much longer..

Reply:
I made it more consistant